src/js/db/DBD.js

```diff
@@ -0,0 +1,179 @@
+const util = require('util');
+const log = require('../log');
+const core = require('../core');
+const generics = require('../generics');
+
+module.exports = {
+	/**
+	 * Generates a DB schema based upon WoWDBDefs.
+	 * @param {string} tableName
+	 * @param {string} layoutHash
+	 * @returns {object, object}
+	 */
+	getSchemaForTable: async (tableName, layouthash) => {
+		// Download DBD for specific table (keep case sensitivity in mind)
+		const url = util.format(core.view.config.dbdURL, tableName)
+		let data = await generics.downloadFile(url);
+
+		const targetBuild = core.view.casc.getBuildName();
+
+		// Parse DBD
+		const columnDefinitions = [];
+		const versionDefinitions = [];
+
+		const validTypes = ["int", "float", "string", "locstring"];
+
+		const lines = data.readLines();
+		let lineNumber = 0;
+
+		if(lines[0].substring(0, 7) === "COLUMNS"){
+			lineNumber++;
+			while(true){
+				const line = lines[lineNumber++];
+
+				// Column definitions end at an empty line
+				if(line.length === 0)
+					break;
+
+				// TYPE READING, find index of space (end of type) or < (foreign key)
+				const spacePos = line.indexOf(' ');
+				const larrPos = line.indexOf('<');
+				const type = line.substring(0, (larrPos != -1 && spacePos > larrPos) ? larrPos : spacePos);
+
+				// Check if type is valid, otherwise error out
+				if(!validTypes.includes(type))
+					throw new Error('Unsupported DBD type: ' + type);
+
+				// SKIPPED: Foreign key reading
+
+				// NAME READING
+				const firstSpace = line.indexOf(' ');
+				const lastSpace = line.lastIndexOf(' ');
+				const name = line.substring(firstSpace + 1, (firstSpace === lastSpace) ? line.length : line.indexOf(' ', firstSpace + 1)).replace('?', '');
+
+				// SKIPPED: Verified (?) and comment reading
+
+				columnDefinitions[name] = type;
+			}
+		}
+
+		let layouthashes = [];
+		let builds = [];
+		let fields = [];
+
+		for(let i = lineNumber; i < lines.length; i++){
+			// Line is manipulated later on, not a const
+			let line = lines[i];
+
+			if(line.length === 0){
+				versionDefinitions.push({fields, builds, layouthashes});
+
+				layouthashes = [];
+				builds = [];
+				fields = [];
+			}
+
+			// Parse layouthashes
+			if(line.substring(0, 6) === "LAYOUT"){
+				layouthashes = line.replace("LAYOUT ", "").split(", ");
+			}
+
+			// Parse builds, skip BUILD lines with ranges
+			if(line.substring(0, 5) === "BUILD" && !line.includes('-')){
+				builds = builds.concat(builds, line.replace("BUILD ", "").split(", "));
+			}
+
+			if (line.substring(0, 6) !== "LAYOUT" && line.substring(0, 5) !== "BUILD" && line.substring(0, 7) !== "COMMENT" && line.length !== 0)
+			{
+				let field = [];
```

Comment:
Array used like associative array in PHP. Better to use an object here. Also not reassigned so can be constant.

`const field = { arrLength = 0 }`

`isID`, `isNonInline` and `isRelation` can be omitted since non-existent object properties default to `undefined` which is falsey.